File: clients/ios/Sources/SpectreUI/Components/Inputs.swift
import SwiftUI
import SpectreCore

struct ButtonView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        let label = node.string("label")
        let loading = node.bool("loading", default: false)
        // loading 中はタップを無効化する。多重発火の防止は ActionDispatcher 側でも
        // 行うが、押せてしまうこと自体が誤操作の原因になるので UI でも止める。
        let enabled = node.bool("enabled", default: true) && !loading
        let variant = node.token("variant", default: "primary")

        Button {
            model.dispatch(node.actions("onTap"))
        } label: {
            HStack(spacing: theme.space("xs")) {
                if loading {
                    ProgressView().controlSize(.small)
                } else if let icon = node.tokenOrNil("leadingIcon") {
                    Image(systemName: theme.symbol(icon))
                }
                if !label.isEmpty { Text(label) }
                if let icon = node.tokenOrNil("trailingIcon") {
                    Image(systemName: theme.symbol(icon))
                }
            }
            .frame(maxWidth: node.layout["width"]?.asString == "fill" ? .infinity : nil)
        }
        .disabled(!enabled)
        .modifier(ButtonVariantModifier(variant: variant, size: node.token("size", default: "md")))
        .spectreNode(node)
    }
}

private struct ButtonVariantModifier: ViewModifier {
    let variant: String
    let size: String

    func body(content: Content) -> some View {
        let controlSize: ControlSize = size == "sm" ? .small : (size == "lg" ? .large : .regular)
        switch variant {
        case "secondary":
            return AnyView(content.buttonStyle(.bordered).controlSize(controlSize))
        case "tertiary", "text":
            return AnyView(content.buttonStyle(.borderless).controlSize(controlSize))
        case "destructive":
            return AnyView(content.buttonStyle(.borderedProminent).tint(.red).controlSize(controlSize))
        default:
            return AnyView(content.buttonStyle(.borderedProminent).controlSize(controlSize))
        }
    }
}

struct TextFieldView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel
    @Environment(\.spectreTheme) private var theme
    @State private var text: String = ""
    @State private var debounceTask: Task<Void, Never>?
    /// `focus` アクション (docs/spec/actions.md) の着地点。
    @FocusState private var isFocused: Bool

    var body: some View {
        let bindTo = node.stringOrNil("bindTo")
        let bound = bindTo.map { model.stateValue($0) } ?? .null
        let boundText = bound.isNull ? "" : bound.stringify()
        let errorText = node.stringOrNil("errorText")
        let keyboard = node.token("keyboard", default: "text")

        VStack(alignment: .leading, spacing: 4) {
            if let label = node.stringOrNil("label") {
                Text(label).font(theme.font("caption")).foregroundStyle(.secondary)
            }

            Group {
                if keyboard == "password" {
                    SecureField(node.stringOrNil("placeholder") ?? "", text: $text)
                } else if node.bool("multiline", default: false) {
                    TextField(node.stringOrNil("placeholder") ?? "", text: $text, axis: .vertical)
                        .lineLimit(3...6)
                } else {
                    TextField(node.stringOrNil("placeholder") ?? "", text: $text)
                }
            }
            .textFieldStyle(.roundedBorder)
            .autocorrectionDisabled(keyboard == "email" || keyboard == "url")
            // キーボード種別と自動大文字化は UIKit 由来で macOS には無い。
            // CI が macOS ホストで `swift build` を通せるよう分岐しておく。
            .modifier(KeyboardTraitsModifier(keyboard: keyboard))
            .focused($isFocused)
            .onSubmit {
                if let bindTo { model.setStateValue(bindTo, .string(text)) }
                model.dispatch(node.actions("onSubmit"))
            }

            if let supporting = errorText ?? node.stringOrNil("helperText") {
                Text(supporting)
                    .font(theme.font("caption"))
                    .foregroundStyle(errorText != nil ? theme.color("error", default: .red) : .secondary)
            }
        }
        // 入力中は自前の状態で追従し、確定後に state へ反映する。
        // state 経由で毎打鍵を往復させるとカーソル位置が飛ぶため。
        .onAppear { text = boundText }
        .onChange(of: boundText) { newValue in
            if newValue != text { text = newValue }
        }
        .onChange(of: model.focusRequest) { newValue in
            guard let nodeID = node.nodeID, newValue == nodeID else { return }
            isFocused = true
            model.consumeFocusRequest()
        }
        .onChange(of: text) { newValue in
            guard let bindTo, newValue != boundText else { return }
            let maxLength = node.intOrNil("maxLength")
            let clipped = maxLength.map { String(newValue.prefix($0)) } ?? newValue
            if clipped != newValue { text = clipped; return }

            let onChange = node.actions("onChange")
            guard !onChange.isEmpty else {
                model.setStateValue(bindTo, .string(clipped))
                return
            }
            // onChange はデバウンス後に発火する。1文字ごとにサーバへ飛ばさないための既定値。
            debounceTask?.cancel()
            let delayMs = node.int("debounceMs", default: 300)
            debounceTask = Task { @MainActor in
                try? await Task.sleep(nanoseconds: UInt64(min(max(delayMs, 0), 5000)) * 1_000_000)
                guard !Task.isCancelled else { return }
                model.setStateValue(bindTo, .string(clipped))
                model.dispatch(onChange)
            }
        }
        .spectreNode(node)
    }
}

/// キーボード種別と自動大文字化の指定。
///
/// どちらも UIKit 由来で macOS には存在しないため、iOS 以外では何もしない。
/// SpectreCore の適合性テストを macOS ホストで走らせる都合上、
/// SpectreUI も macOS でコンパイルが通る必要がある。
private struct KeyboardTraitsModifier: ViewModifier {
    let keyboard: String

    func body(content: Content) -> some View {
        #if os(iOS)
        content
            .keyboardType(keyboardType)
            .textInputAutocapitalization(keyboard == "email" || keyboard == "url" ? .never : .sentences)
        #else
        content
        #endif
    }

    #if os(iOS)
    private var keyboardType: UIKeyboardType {
        switch keyboard {
        case "email": return .emailAddress
        case "number": return .numberPad
        case "phone": return .phonePad
        case "url": return .URL
        default: return .default
        }
    }
    #endif
}

struct ToggleView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            Toggle(node.stringOrNil("label") ?? "", isOn: Binding(
                get: { model.stateValue(bindTo).isTruthy },
                set: { newValue in
                    model.setStateValue(bindTo, .bool(newValue))
                    model.dispatch(node.actions("onChange"))
                }
            ))
            .disabled(!node.bool("enabled", default: true))
            .spectreNode(node)
        }
    }
}

/// SwiftUI に Checkbox はないため、トグル可能な行として表現する。
/// iOS の慣習に合わせつつ、Android の Checkbox と同じ操作結果になるようにしている。
struct CheckboxView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            let checked = model.stateValue(bindTo).isTruthy
            Button {
                model.setStateValue(bindTo, .bool(!checked))
                model.dispatch(node.actions("onChange"))
            } label: {
                HStack(spacing: theme.space("sm")) {
                    Image(systemName: checked ? "checkmark.square.fill" : "square")
                        .foregroundStyle(checked ? theme.color("primary", default: .accentColor) : .secondary)
                    if let label = node.stringOrNil("label") {
                        Text(label).foregroundStyle(.primary)
                    }
                    Spacer(minLength: 0)
                }
            }
            .buttonStyle(.plain)
            .disabled(!node.bool("enabled", default: true))
            .accessibilityAddTraits(checked ? [.isButton, .isSelected] : .isButton)
            .spectreNode(node)
        }
    }
}

struct RadioGroupView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel
    @Environment(\.spectreTheme) private var theme

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            let selected = model.stateValue(bindTo).stringify()
            let options = node.options("options")
            let horizontal = node.token("orientation", default: "vertical") == "horizontal"

            Group {
                if horizontal {
                    HStack(spacing: theme.space("sm")) {
                        ForEach(options) { option in row(option, selected: selected, bindTo: bindTo) }
                    }
                } else {
                    VStack(alignment: .leading, spacing: theme.space("xs")) {
                        ForEach(options) { option in row(option, selected: selected, bindTo: bindTo) }
                    }
                }
            }
            .spectreNode(node)
        }
    }

    @ViewBuilder
    private func row(_ option: SpectreOption, selected: String, bindTo: String) -> some View {
        let isSelected = option.value == selected
        Button {
            model.setStateValue(bindTo, .string(option.value))
            model.dispatch(node.actions("onChange"))
        } label: {
            HStack(spacing: 6) {
                Image(systemName: isSelected ? "largecircle.fill.circle" : "circle")
                    .foregroundStyle(isSelected ? theme.color("primary", default: .accentColor) : .secondary)
                Text(option.label).foregroundStyle(.primary)
            }
        }
        .buttonStyle(.plain)
        .disabled(!option.enabled)
        .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
    }
}

struct SelectView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            let options = node.options("options")
            let selected = model.stateValue(bindTo).stringify()

            Picker(node.stringOrNil("label") ?? "", selection: Binding(
                get: { selected },
                set: { newValue in
                    model.setStateValue(bindTo, .string(newValue))
                    model.dispatch(node.actions("onChange"))
                }
            )) {
                if !options.contains(where: { $0.value == selected }) {
                    Text(node.stringOrNil("placeholder") ?? "選択してください").tag(selected)
                }
                ForEach(options) { option in
                    Text(option.label).tag(option.value)
                }
            }
            // searchable が真ならシート表示に切り替える想定だが、v0.1 はメニューのみ。
            .pickerStyle(.menu)
            .spectreNode(node)
        }
    }
}

struct SliderView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            let minValue = node.double("min", default: 0)
            let maxValue = node.double("max", default: 100)
            let step = node.double("step", default: 1)
            let current = model.stateValue(bindTo).asDouble ?? minValue

            VStack(alignment: .leading, spacing: 4) {
                if node.bool("showValue", default: false) {
                    Text(SpValue.formatNumberPlain(current))
                }
                Slider(
                    value: Binding(
                        get: { min(max(current, minValue), maxValue) },
                        set: { model.setStateValue(bindTo, .number($0)) }
                    ),
                    in: minValue...max(maxValue, minValue + 1),
                    step: step > 0 ? step : 1,
                    onEditingChanged: { editing in
                        if !editing { model.dispatch(node.actions("onChange")) }
                    }
                )
            }
            .spectreNode(node)
        }
    }
}

struct StepperView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            let step = node.double("step", default: 1)
            let minValue = node.doubleOrNil("min") ?? 0
            let maxValue = node.doubleOrNil("max")
            let current = model.stateValue(bindTo).asDouble ?? minValue

            Stepper {
                Text(SpValue.formatNumberPlain(current))
            } onIncrement: {
                let next = current + step
                guard maxValue == nil || next <= maxValue! else { return }
                model.setStateValue(bindTo, .number(next))
                model.dispatch(node.actions("onChange"))
            } onDecrement: {
                let next = current - step
                guard next >= minValue else { return }
                model.setStateValue(bindTo, .number(next))
                model.dispatch(node.actions("onChange"))
            }
            .spectreNode(node)
        }
    }
}

struct DatePickerView: View {
    let node: RenderNode
    @EnvironmentObject private var model: SpectreScreenModel

    /// IconView と同じ理由で body の外に出す。@ViewBuilder の下では
    /// 代入の switch が View 式として解釈される。
    private var components: DatePickerComponents {
        switch node.token("mode", default: "date") {
        case "time": return [.hourAndMinute]
        case "dateTime": return [.date, .hourAndMinute]
        default: return [.date]
        }
    }

    var body: some View {
        if let bindTo = node.stringOrNil("bindTo") {
            let raw = model.stateValue(bindTo).stringify()
            DatePicker(
                node.stringOrNil("label") ?? "",
                selection: Binding(
                    get: { SpectreDate.parse(raw) ?? Date() },
                    set: { newDate in
                        model.setStateValue(bindTo, .string(SpectreDate.format(newDate)))
                        model.dispatch(node.actions("onChange"))
                    }
                ),
                displayedComponents: components
            )
            .spectreNode(node)
        }
    }
}

/// ドキュメント側の日付表現は常に ISO 8601 文字列。
/// Android 側も同じ文字列を state に置くため、サーバから見て差が出ない。
enum SpectreDate {
    private static let formatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func parse(_ text: String) -> Date? {
        guard !text.isEmpty else { return nil }
        if let date = formatter.date(from: text) { return date }
        // "2026-03-05" のような日付のみの表記も受ける
        let dateOnly = DateFormatter()
        dateOnly.calendar = Calendar(identifier: .gregorian)
        dateOnly.locale = Locale(identifier: "en_US_POSIX")
        dateOnly.timeZone = TimeZone(identifier: "UTC")
        dateOnly.dateFormat = "yyyy-MM-dd"
        return dateOnly.date(from: text)
    }

    static func format(_ date: Date) -> String { formatter.string(from: date) }
}

File: clients/android/spectre-ui/src/main/kotlin/dev/spectre/ui/components/Inputs.kt
package dev.spectre.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Slider
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import dev.spectre.core.RenderNode
import dev.spectre.core.SpValue
import dev.spectre.core.asDoubleOrNull
import dev.spectre.core.isTruthy
import dev.spectre.core.stringify
import dev.spectre.ui.LocalSpectreController
import dev.spectre.ui.LocalSpectreTheme
import dev.spectre.ui.SpectreOption
import dev.spectre.ui.a11yLabel
import dev.spectre.ui.bool
import dev.spectre.ui.float
import dev.spectre.ui.floatOrNull
import dev.spectre.ui.int
import dev.spectre.ui.intOrNull
import dev.spectre.ui.options
import dev.spectre.ui.spectreNode
import dev.spectre.ui.string
import dev.spectre.ui.stringOrNull
import dev.spectre.ui.token

@Composable
fun ButtonView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val controller = LocalSpectreController.current
    val label = node.string("label")
    val loading = node.bool("loading", false)
    // loading 中はタップを無効化する。多重発火の防止は ActionDispatcher 側でも行うが、
    // 押せてしまうこと自体が誤操作の原因になるので UI でも止める。
    val enabled = node.bool("enabled", true) && !loading
    val onTap = { controller.dispatch(node.actions("onTap")) }
    val variant = node.token("variant", "primary")

    val content: @Composable () -> Unit = {
        Row(
            horizontalArrangement = Arrangement.spacedBy(theme.space("xs")),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (loading) {
                CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp)
            } else {
                node.stringOrNull("leadingIcon")?.let {
                    Icon(theme.icon(it), contentDescription = null, modifier = Modifier.size(18.dp))
                }
            }
            if (label.isNotEmpty()) Text(label)
            node.stringOrNull("trailingIcon")?.let {
                Icon(theme.icon(it), contentDescription = null, modifier = Modifier.size(18.dp))
            }
        }
    }

    val base = modifier.spectreNode(node).applyA11y(node)

    when (variant) {
        "secondary" -> OutlinedButton(onTap, base, enabled = enabled) { content() }
        "tertiary", "text" -> TextButton(onTap, base, enabled = enabled) { content() }
        "destructive" -> Button(
            onTap,
            base,
            enabled = enabled,
            colors = ButtonDefaults.buttonColors(
                containerColor = theme.color("error", MaterialTheme.colorScheme.error),
                contentColor = theme.color("onError", MaterialTheme.colorScheme.onError),
            ),
        ) { content() }
        else -> {
            // ラベルが空でアイコンだけのボタンは IconButton に落とす。
            // a11y.label はリントで必須になっているため、ここでは指定を反映するだけ。
            val iconOnly = label.isEmpty() && node.stringOrNull("leadingIcon") != null
            if (iconOnly) {
                IconButton(onTap, base, enabled = enabled) {
                    Icon(theme.icon(node.stringOrNull("leadingIcon")), contentDescription = node.a11yLabel())
                }
            } else {
                Button(onTap, base, enabled = enabled) { content() }
            }
        }
    }
}

@Composable
fun TextFieldView(node: RenderNode, modifier: Modifier) {
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val bound = controller.stateValue(bindTo).let { if (it is SpValue.Null) "" else it.stringify() }

    // 入力中は自前の状態で追従し、確定後に state へ反映する。
    // state 経由だと再解決のたびにカーソル位置が飛ぶため。
    var text by remember(bindTo) { mutableStateOf(bound) }
    LaunchedEffect(bound) { if (bound != text) text = bound }

    val debounceMs = node.int("debounceMs", 300)
    val onChange = node.actions("onChange")

    // onChange はデバウンス後に発火する。1文字ごとにサーバへ飛ばさないための既定値。
    LaunchedEffect(text, onChange.isNotEmpty()) {
        if (onChange.isEmpty() || text == bound) return@LaunchedEffect
        kotlinx.coroutines.delay(debounceMs.toLong().coerceIn(0, 5_000))
        controller.setStateValue(bindTo, SpValue.Str(text))
        controller.dispatch(onChange)
    }

    val keyboard = node.token("keyboard", "text")
    val isPassword = keyboard == "password"
    val errorText = node.stringOrNull("errorText")

    OutlinedTextField(
        value = text,
        onValueChange = { next ->
            val maxLength = node.intOrNull("maxLength")
            text = if (maxLength != null && next.length > maxLength) next.take(maxLength) else next
            if (onChange.isEmpty()) controller.setStateValue(bindTo, SpValue.Str(text))
        },
        modifier = modifier.spectreNode(node).fillMaxWidth(),
        label = node.stringOrNull("label")?.let { { Text(it) } },
        placeholder = node.stringOrNull("placeholder")?.let { { Text(it) } },
        supportingText = (errorText ?: node.stringOrNull("helperText"))?.let { { Text(it) } },
        isError = errorText != null,
        singleLine = !node.bool("multiline", false),
        visualTransformation = if (isPassword) PasswordVisualTransformation() else VisualTransformation.None,
        keyboardOptions = KeyboardOptions(
            keyboardType = when (keyboard) {
                "email" -> KeyboardType.Email
                "number" -> KeyboardType.Number
                "phone" -> KeyboardType.Phone
                "url" -> KeyboardType.Uri
                "password" -> KeyboardType.Password
                else -> KeyboardType.Text
            },
            imeAction = if (node.actions("onSubmit").isNotEmpty()) ImeAction.Done else ImeAction.Default,
        ),
        keyboardActions = KeyboardActions(
            onDone = {
                controller.setStateValue(bindTo, SpValue.Str(text))
                controller.dispatch(node.actions("onSubmit"))
            }
        ),
    )
}

@Composable
fun ToggleView(node: RenderNode, modifier: Modifier) {
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val checked = controller.stateValue(bindTo).isTruthy

    Row(
        modifier = modifier.spectreNode(node).fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        node.stringOrNull("label")?.let { Text(it) }
        Switch(
            checked = checked,
            onCheckedChange = {
                controller.setStateValue(bindTo, SpValue.Bool(it))
                controller.dispatch(node.actions("onChange"))
            },
            enabled = node.bool("enabled", true),
        )
    }
}

@Composable
fun CheckboxView(node: RenderNode, modifier: Modifier) {
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val checked = controller.stateValue(bindTo).isTruthy
    val enabled = node.bool("enabled", true)

    val toggle = {
        controller.setStateValue(bindTo, SpValue.Bool(!checked))
        controller.dispatch(node.actions("onChange"))
    }

    Row(
        modifier = modifier.spectreNode(node),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Checkbox(checked = checked, onCheckedChange = { toggle() }, enabled = enabled)
        node.stringOrNull("label")?.let {
            Text(it, modifier = Modifier.padding(start = 4.dp))
        }
    }
}

@Composable
fun RadioGroupView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val selected = controller.stateValue(bindTo).let { if (it is SpValue.Null) null else it.stringify() }
    val opts = node.options("options")

    val select: (SpectreOption) -> Unit = { option ->
        controller.setStateValue(bindTo, SpValue.Str(option.value))
        controller.dispatch(node.actions("onChange"))
    }

    val item: @Composable (SpectreOption) -> Unit = { option ->
        Row(
            modifier = Modifier
                .selectable(
                    selected = option.value == selected,
                    enabled = option.enabled,
                    onClick = { select(option) },
                ),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            RadioButton(
                selected = option.value == selected,
                onClick = { select(option) },
                enabled = option.enabled,
            )
            Text(option.label)
        }
    }

    if (node.token("orientation", "vertical") == "horizontal") {
        Row(
            modifier = modifier.spectreNode(node),
            horizontalArrangement = Arrangement.spacedBy(theme.space("sm")),
        ) { opts.forEach { item(it) } }
    } else {
        Column(modifier = modifier.spectreNode(node)) { opts.forEach { item(it) } }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SelectView(node: RenderNode, modifier: Modifier) {
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val opts = node.options("options")
    val selected = controller.stateValue(bindTo).let { if (it is SpValue.Null) null else it.stringify() }
    val selectedLabel = opts.firstOrNull { it.value == selected }?.label
        ?: node.stringOrNull("placeholder")
        ?: ""

    var expanded by remember { mutableStateOf(false) }

    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = it },
        modifier = modifier.spectreNode(node),
    ) {
        OutlinedTextField(
            value = selectedLabel,
            onValueChange = {},
            readOnly = true,
            label = node.stringOrNull("label")?.let { { Text(it) } },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            modifier = Modifier
                .menuAnchor(androidx.compose.material3.MenuAnchorType.PrimaryNotEditable)
                .fillMaxWidth(),
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            opts.forEach { option ->
                DropdownMenuItem(
                    text = { Text(option.label) },
                    enabled = option.enabled,
                    onClick = {
                        expanded = false
                        controller.setStateValue(bindTo, SpValue.Str(option.value))
                        controller.dispatch(node.actions("onChange"))
                    },
                )
            }
        }
    }
}

@Composable
fun SliderView(node: RenderNode, modifier: Modifier) {
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val min = node.float("min", 0f)
    val max = node.float("max", 100f)
    val step = node.float("step", 1f)
    val current = controller.stateValue(bindTo).asDoubleOrNull?.toFloat() ?: min

    // Compose の steps は「両端を除いた刻みの数」
    val steps = if (step > 0f && max > min) (((max - min) / step).toInt() - 1).coerceAtLeast(0) else 0

    Column(modifier.spectreNode(node)) {
        if (node.bool("showValue", false)) Text(current.toInt().toString())
        Slider(
            value = current.coerceIn(min, max),
            onValueChange = { controller.setStateValue(bindTo, SpValue.Num(it.toDouble())) },
            onValueChangeFinished = { controller.dispatch(node.actions("onChange")) },
            valueRange = min..max,
            steps = steps,
        )
    }
}

@Composable
fun StepperView(node: RenderNode, modifier: Modifier) {
    val theme = LocalSpectreTheme.current
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val step = node.float("step", 1f)
    val min = node.floatOrNull("min") ?: 0f
    val max = node.floatOrNull("max")
    val current = controller.stateValue(bindTo).asDoubleOrNull?.toFloat() ?: min

    val apply: (Float) -> Unit = { next ->
        val clamped = if (max != null) next.coerceIn(min, max) else next.coerceAtLeast(min)
        if (clamped != current) {
            controller.setStateValue(bindTo, SpValue.Num(clamped.toDouble()))
            controller.dispatch(node.actions("onChange"))
        }
    }

    Row(
        modifier = modifier.spectreNode(node),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(theme.space("xs")),
    ) {
        // マイナスは Material のコアアイコン集合にないため文字で描く
        TextButton(
            onClick = { apply(current - step) },
            enabled = current > min,
        ) { Text("−") }

        Text(
            text = current.toInt().toString(),
            style = MaterialTheme.typography.titleMedium,
        )

        TextButton(
            onClick = { apply(current + step) },
            enabled = max == null || current < max,
        ) { Text("+") }
    }
}

/**
 * 日付入力。
 *
 * v0.1 ではカレンダーダイアログを出さず、ISO 8601 文字列の直接入力にしている。
 * Material3 の DatePickerDialog は実験的 API で、SwiftUI 側の DatePicker と
 * 挙動を揃えるのに追加の設計が要るため、先に他を固めてから取り組む。
 */
@Composable
fun DatePickerView(node: RenderNode, modifier: Modifier) {
    val controller = LocalSpectreController.current
    val bindTo = node.stringOrNull("bindTo") ?: return
    val value = controller.stateValue(bindTo).let { if (it is SpValue.Null) "" else it.stringify() }

    OutlinedTextField(
        value = value,
        onValueChange = {
            controller.setStateValue(bindTo, SpValue.Str(it))
            controller.dispatch(node.actions("onChange"))
        },
        modifier = modifier.spectreNode(node).fillMaxWidth(),
        label = node.stringOrNull("label")?.let { { Text(it) } },
        placeholder = { Text("YYYY-MM-DD") },
        singleLine = true,
    )
}
